src/os/bluestore/BlueStore.cc

```diff
@@ -3731,14 +3731,13 @@ void BlueStore::Collection::split_cache(
       ldout(store->cct, 20) << __func__ << " moving " << o << " " << o->oid
 			    << dendl;
 
-      onode_map.cache->_rm(p->second);
       p = onode_map.onode_map.erase(p);
-
       o->c = dest;
-      dest->onode_map.cache->_add(o, 1);
       dest->onode_map.onode_map[o->oid] = o;
-      dest->onode_map.cache = dest->onode_map.cache;
-
+      if (dest->cache != cache) {
```

Comment:
IMO this check is a bit confusing, proper one should rather be:
if (onode_map.cache != dest->onode_map.cache)
...

Reply:
Hi Igor, Josh and I spent some time in gdb looking at the state of the ref counting. I don't remember exactly everything we determined, but I believe we convinced ourselves that this was indeed due to issues with the ref counting (specifically due to the new implementation of the cache triming).  Given that Kefu was able to hit it again though, it appears to happen in general rather than specific to situations where both of the resulting collections are in the same cache shard.

Reply:
@ifed01 fwiw, I agree on the check, though I think it's currently working by accident (the buffer cache check should hash out the same way).  I don't think this is what's causing the assert kefu saw.  For some reason the cache shard is not null when _add is called, but I haven't been able to figure out why yet.